pest/examples/parens.rs

```diff
@@ -0,0 +1,105 @@
+extern crate pest;
+
+use std::io::{self, Write};
+use std::rc::Rc;
+
+use pest::inputs::{Input, Position, StringInput};
+use pest::iterators::Pairs;
+use pest::{Error, Parser, ParserState, state};
+
+#[allow(non_camel_case_types)]
+#[allow(dead_code)]
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+enum Rule {
+    expr,
+    paren,
+    paren_end
+}
+
+struct ParenParser;
+
+impl Parser<Rule> for ParenParser {
+    fn parse<I: Input>(rule: Rule, input: Rc<I>) -> Result<Pairs<Rule, I>, Error<Rule, I>> {
+        fn expr<I: Input>(
```

Comment:
Why are these functions defined inside `parse()`? Would it be possible to avoid the extra nesting and move them out? They don't seem to be benefiting from the nesting. Though they are hidden from outside the function, there are other ways to accomplish that without making this function so large.